datafusion/core/src/physical_plan/union.rs

```diff
@@ -247,6 +247,30 @@ impl ExecutionPlan for UnionExec {
         }
     }
 
+    fn maintains_input_order(&self) -> bool {
+        let first_input_ordering = self.inputs[0].output_ordering();
```

Comment:
This is basically exactly what was suggested by @mustafasrepo  with a slightly modified comment